---
layout: post
title:  "HW2: Spatial gene expression with PCA"
author: Suki Ogihara
jhed: iogihar1
categories: [ HW2 ]
image: homework/hw2/itsuki_ogihara.png
featured: true
---

### 1. What data types are you visualizing?
* Spatial data of each cell, the x, y  coordinates of the cell location.
* Quantitative: each dot is color coded with the 1st principal component, PC1 as well as PC2. The color is mapped to the value of the PC1 and PC2.
* Same process was repeated using normalized data, shown in the bottom row. Each row is normalized by the total count of all genes in each cell as a proxy to account for the cell size and densiy.

### 2. What data encodings (geometric primitives and visual channels) are you using to visualize these data types?
* Used the points as geometric primitive for each cell positioned by x and y coordinates. 
* Used color to encode the values of PC1 and PC2 respectively. Note that their ranges are set to the same range to make the color mapping consistent across all plots for better comparison.

### 3. What about the data are you trying to make salient through this data visualization? 
The intention is to use the PCA to reduce the dimensionality of the data and visualize the relationship between overall gene expression level vs spatial distribution of the cells. PCA allows us to have the compact representation of the data and visualize the values in reduced space. 

### 4. What Gestalt principles or knowledge about perceptiveness of visual encodings are you using to accomplish this?
* Similarity: cells with similar values of PC1 and PC2 are represented by similar colors. 
* Proximity of the dots to each other shows the spatial distribution of the gene across the cells.
* As the plot shows indeed the cells with similar gene expression levels are located close to each other and follow the spatial distribution of the cells.

### 5. Code (paste your code in between the ``` symbols)

```r
file <- '~/Desktop/genomic-data-visualization-2025/data/pikachu.csv.gz'
data <- read.csv(file, row.names=1)

library(gridExtra)
library(ggplot2)
library(stats)

dir <-"/Users/suki/Projects/genomic-data-visualization-2025/"
data <- read.csv("/Users/suki/Projects/genomic-data-visualization-2025//data/pikachu.csv.gz")

gexp <- data[, 7:ncol(data)]
normed_gexp <- gexp/rowSums(gexp)

# PCA 
pca <- prcomp(gexp, scale=TRUE)
normed_pca <- prcomp(normed_gexp, scale=TRUE)

df <- data.frame(pca$x)
normed_df <- data.frame(normed_pca$x)
data_pca <- cbind(data, df)
data_normed_pca <- cbind(data, normed_df)

grid.arrange(
  ggplot(data_pca) + geom_point(aes(x=aligned_x, y=aligned_y, color=PC1), size=0.5) + ggtitle("Spatial Distribution of PC1") + theme(aspect.ratio=1.0) + scale_color_gradientn(colors = rainbow(10), limits=c(-10, 40)) + xlab("Aligned X") + ylab("Aligned Y") + theme(text = element_text(size=15)),
  ggplot(data_pca) + geom_point(aes(x=aligned_x, y=aligned_y, color=PC2), size=0.5) + ggtitle("Spatial Distribution of PC2") + theme(aspect.ratio=1.0) + scale_color_gradientn(colors = rainbow(10), limits=c(-10, 40)) + xlab("Aligned X") + ylab("Aligned Y") + theme(text = element_text(size=15)),
  ggplot(data_normed_pca) + geom_point(aes(x=aligned_x, y=aligned_y, color=PC1), size=0.5) + ggtitle("Spatial Distribution of PC1(Normalized)") + theme(aspect.ratio=1.0) + scale_color_gradientn(colors = rainbow(10), limits=c(-10, 40)) + xlab("Aligned X") + ylab("Aligned Y") + theme(text = element_text(size=15)),
  ggplot(data_normed_pca) + geom_point(aes(x=aligned_x, y=aligned_y, color=PC2), size=0.5) + ggtitle("Spatial Distribution of PC2 (Normalized)") + theme(aspect.ratio=1.0) + scale_color_gradientn(colors = rainbow(10), limits=c(-10, 40)) + xlab("Aligned X") + ylab("Aligned Y") + theme(text = element_text(size=15)),
  ncol = 2
)

```