---
layout: page
title:  "Lesson 11: Spatially variable genes"
order: 11
---

# Lecture

## Learning objectives

By the end of this lesson, we should be able to understand how to identify spatially variable genes in our data using Moran's I. We will also be able to use interactive data visualizations to get some sense for the quality of our analysis.

---

# Hands-on Component

Our in-class hands-on component will involve calculating Moran's I. 

---

# Class Notes

Prof. Fan’s notes from class: [genomic-data-visualization-Lesson_11.pptx (click to download)]({{ site.baseurl }}/resources/notes/genomic-data-visualization-Lesson_11.pptx)

Prof. Fan’s code from class: [code-lesson-11.R (click to download)]({{ site.baseurl }}/resources/code/code-lesson-11.R)

---

# Homework Assignment

`HW_4` apply same approach for analyzing your original spatial transcriptomics dataset to the other team's dataset for comparison purposes. 

See slides for more details on what is expected: [genomic-data-visualization-HW_4.pptx (click to download)]({{ site.baseurl }}/resources/hw/genomic-data-visualization-HW_4.pptx)

---

# Additional resources
- https://jef.works/blog/2024/08/29/the-many-ways-to-detect-svgs-using-moransI/
